File: src/trigger_arc_tsp/instance.py
from __future__ import annotations

import os
import pickle
from collections import defaultdict
from datetime import datetime
from warnings import warn

from trigger_arc_tsp.gurobi_tsp_model import GurobiTSPModel
from trigger_arc_tsp.utils import CACHE_DIR, SOLUTIONS_DIR


class Instance:
    def __init__(self, N: int, edges: dict, relations: dict, name: str) -> None:
        self.name = name
        self.model_name = name.removesuffix(".txt") + ".mps"
        self.cache_file = os.path.join(
            CACHE_DIR, name.removesuffix(".txt") + "indices_cache.pkl"
        )

        os.makedirs(CACHE_DIR, exist_ok=True)
        os.makedirs(os.path.dirname(self.cache_file), exist_ok=True)

        self.edges = edges
        self.relations = relations

        self.N = N
        self.nodes = set(range(self.N))
        self.A = len(edges)
        self.R = len(self.relations)

        self.R_a = defaultdict(list)
        for b_0, b_1, a_0, a_1 in self.relations:
            self.R_a[(a_0, a_1)].append((b_0, b_1))

        self.delta_in = {node: set() for node in self.nodes}
        self.delta_out = {node: set() for node in self.nodes}
        for i, j in edges:
            self.delta_in[j].add(i)
            self.delta_out[i].add(j)

        self.relations = {
            rel: self.relations[rel] - self.edges[(rel[2], rel[3])] for rel in self.relations
        }

        if not self._load_indices_from_cache():
            self._generate_z_var_indices()

    def _generate_z_var_indices(self) -> None:
        # print("Instance loading: Generating indices")
        # Optimize z_var_indices creation
        z_var_indices = set()

        for a, b_list in self.R_a.items():
            for b in b_list:
                if b != 0:
                    z_var_indices.add((*a, *b))  # (*a, *b)
                    z_var_indices.add((*b, *a))  # (*b, *a)
                for c in b_list:
                    if b != c and b not in (c, 0):
                        z_var_indices.add((*b, *c))  # (*b, *c)

        self.z_var_indices = list(z_var_indices)
        self._store_indices_in_cache()

    def _store_indices_in_cache(self) -> None:
        """Stores z_var_indices in a cache file using pickle."""
        # print("Instance loading: Storing indices in cache")
        if self.name == "test" or "tsp" in self.name:
            return
        with open(self.cache_file, "wb") as cache:
            pickle.dump(self.z_var_indices, cache)

    def _load_indices_from_cache(self) -> bool:
        """Loads z_var_indices from the cache file if it exists."""
        if self.name == "test" or "tsp" in self.name:
            return False

        if os.path.exists(self.cache_file):
            with open(self.cache_file, "rb") as cache:
                self.z_var_indices = pickle.load(cache)  # noqa: S301
                return True
        return False

    @staticmethod
    def load_instance_from_file(file_path: os.PathLike) -> Instance:
        name = "/".join(file_path.split("/")[-2:])
        print("reading instance file")
        with open(file_path) as file:
            lines = file.readlines()
            N, A, R = map(int, lines[0].split())

            edges = {}
            for line in lines[1 : 1 + A]:
                _, i, j, w = map(float, line.split())
                edges[(int(i), int(j))] = w
            # assert _ == A - 1

            relations = {}
            for line in lines[1 + A : 1 + A + R]:
                _, _, from_trigger, to_trigger, _, from_arc, to_arc, cost = map(
                    float, line.split()
                )
                relations[(int(from_trigger), int(to_trigger), int(from_arc), int(to_arc))] = (
                    cost
                )

            assert int(line.split(" ")[0]) == R - 1

            assert len(edges) == A
            assert len(relations) == R

        print("initialize Instance object")
        return Instance(N, edges, relations, name)

    def compute_objective(self, tour: list) -> float:
        path = [(tour[i], tour[i + 1] if i < self.N - 1 else 0) for i in range(self.N)]
        path_set = set(path)
        assert len(path) == self.N
        assert path[0][0] == 0
        assert path[-1][1] == 0

        cost = sum(self.edges[a] for a in path)

        for a in path:
            if not self.R_a.get(a):
                continue
            # find the relations that could trigger the arc a
            triggering = set(self.R_a[a]).intersection(path_set)
            if not triggering:
                continue
            # sort the triggering relations by their index in the path (higher index last)
            triggering = sorted(triggering, key=lambda x: path.index(x))
            # remove the triggering arcs that happen after the arc a
            triggering = [rel for rel in triggering if path.index(rel) < path.index(a)]
            if not triggering:
                continue
            # add the relative cost of the triggering relation to the cost of the arc a
            cost += self.relations[*triggering[-1], *a]

        return cost

    def check_solution_correctness(self, tour: list) -> bool:
        if tour[-1] == 0:
            tour = tour[:-1]

        if (
            len(tour) != self.N
            or len(set(tour)) != self.N
            or not set(tour).issubset(self.nodes)
        ):
            return False

        if tour[0] != 0:
            return False

        return True

    def test_solution(self, tour: list, proposed_objective: float) -> bool:
        if not self.check_solution_correctness(tour):
            return False

        cost = self.compute_objective(tour)

        return abs(cost - proposed_objective) < 1e-6

    def save_solution(self, tour: list, objective: float | None = None) -> None:
        if not self.check_solution_correctness(tour):
            msg = f"Solution {tour} is not a valid solution for instance {self.name}"
            raise ValueError(msg)

        if objective is None:
            objective = self.compute_objective(tour)
        if not self.test_solution(tour, objective):
            msg = (
                f"Solution {tour} does not have the correct objective value for instance "
                f"{self.name}"
            )
            msg += f" (computed: {self.compute_objective(tour)}, proposed: {objective})"
            warn(msg, stacklevel=1)
            objective = self.compute_objective(tour)

        timestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")  # noqa: DTZ005
        os.makedirs(os.path.join(SOLUTIONS_DIR, *self.name.split("/")[:-1]), exist_ok=True)

        with open(os.path.join(SOLUTIONS_DIR, *self.name.split("/")), "a") as file:
            if file.tell() != 0:
                file.write("\n")
            assert tour[0] == 0, "Tour should start at 0"
            assert tour[-1] != 0, "Tour should not end at 0"
            file.write(",".join(map(str, tour)))
            file.write(f" | {objective} | {timestamp}")

    def __str__(self) -> str:
        return f"Instance(N={self.N}, A={self.A}, R={self.R})"

    def get_best_known_solution(self, idx: int = 0) -> list | None:
        filename = os.path.join(SOLUTIONS_DIR, *self.name.split("/"))
        if not os.path.exists(filename):
            return None

        sols = []
        with open(filename) as file:
            lines = file.readlines()
            for line in lines:
                tour, obj, _ = line.split(" | ")
                if tour not in {s[0] for s in sols}:
                    sols.append((tour, float(obj)))

        sols = sorted(sols, key=lambda x: x[1])
        if idx >= len(sols):
            error = f"Trying to get solution {idx} from {len(sols)} solutions"
            raise ValueError(error)
        tour = list(map(int, sols[idx][0].split(",")))
        print(f"Best known solution for {self.name}: {tour} with objective {sols[idx][1]}")

        return tour

    def get_mip_start(self, *, use_tsp_only: bool = False) -> list[dict, dict, dict, dict]:
        best_known_solution = self.get_best_known_solution()
        tour = (
            best_known_solution
            if best_known_solution and not use_tsp_only
            else self.tsp_solution()
        )
        return self.get_variables_from_tour(tour)

    def get_variables_from_tour(self, tour: list) -> list[dict, dict, dict, dict]:
        tour_edges = [(tour[i], tour[i + 1] if i < self.N - 1 else 0) for i in range(self.N)]
        assert len(tour_edges) == self.N
        x = {edge: 1 if edge in tour_edges else 0 for edge in self.edges}
        u = {node: i for i, node in enumerate(tour)}
        y = dict.fromkeys(self.relations, 0)
        for a in tour_edges:
            assert u[a[0]] < u[a[1]] or a[1] == 0
            if a not in self.R_a:
                continue
            # find the relations that could trigger the arc a
            triggering = set(self.R_a[a]).intersection(tour_edges)
            if not triggering:
                continue
            # sort the triggering relations by their index in the path (higher index last)
            triggering = sorted(triggering, key=lambda x: tour_edges.index(x))
            # remove the triggering arcs that happen after the arc a
            triggering = [
                rel for rel in triggering if tour_edges.index(rel) < tour_edges.index(a)
            ]
            if not triggering:
                continue
            # the last relation in the list is the one that triggers the arc a
            y[*triggering[-1], *a] = 1

        z = {
            (a_1, a_2, b_1, b_2): u[a_1] + 1 <= u[b_1]
            for a_1, a_2, b_1, b_2 in self.z_var_indices
        }

        return x, y, u, z

    def tsp_solution(self) -> list:
        model = GurobiTSPModel(self)
        model.formulate()
        model.solve_to_feasible_solution()
        return model.get_best_tour()

File: cpp_src/src/instance.cpp
#include "instance.hpp"

#include <algorithm>
#include <boost/unordered_map.hpp>
#include <boost/unordered_set.hpp>
#include <chrono>
#include <ctime>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <tsp_model.hpp>

namespace fs = std::filesystem;

Instance::Instance(int N,
                   const boost::unordered_map<std::pair<int, int>, double>& edges,
                   const boost::unordered_map<std::tuple<int, int, int, int>, double>& relations,
                   const std::string& name)
    : name_(name),
      modelName_(name.substr(0, name.find_last_of('.')) + ".mps"),
      N_(N),
      A_(edges.size()),
      R_(relations.size()),
      edges_(edges),
      relations_(relations) {
    // Initialize delta_in and delta_out
    for (int node = 0; node < N_; ++node) {
        deltaIn_[node] = boost::unordered_set<int>();
        deltaOut_[node] = boost::unordered_set<int>();
    }

    // Fill delta_in and delta_out
    for (const auto& [edge, _] : edges_) {
        deltaIn_[edge.second].insert(edge.first);
        deltaOut_[edge.first].insert(edge.second);
    }

    // Initialize R_a
    for (const auto& [rel, cost] : relations_) {
        auto [b0, b1, a0, a1] = rel;
        R_a_[{a0, a1}].push_back({b0, b1});
    }

    // Update relations by subtracting edge costs
    for (auto& [rel, cost] : relations_) {
        auto [b0, b1, a0, a1] = rel;
        cost -= edges_[{a0, a1}];
    }
}

std::unique_ptr<Instance> Instance::loadInstanceFromFile(const std::string& filePath) {
    std::ifstream file(filePath);
    if (!file.is_open()) {
        throw std::runtime_error("Could not open file: " + filePath);
    }

    std::string line;
    std::getline(file, line);
    std::istringstream iss(line);
    int N, A, R;
    iss >> N >> A >> R;

    boost::unordered_map<std::pair<int, int>, double> edges;
    for (int i = 0; i < A; ++i) {
        std::getline(file, line);
        std::istringstream iss(line);
        int idx, from, to;
        double weight;
        iss >> idx >> from >> to >> weight;
        edges[{from, to}] = weight;
    }

    boost::unordered_map<std::tuple<int, int, int, int>, double> relations;
    for (int i = 0; i < R; ++i) {
        std::getline(file, line);
        std::istringstream iss(line);
        int idx1, idx2, fromTrigger, toTrigger, idx3, fromArc, toArc;
        double cost;
        iss >> idx1 >> idx2 >> fromTrigger >> toTrigger >> idx3 >> fromArc >> toArc >> cost;
        relations[{fromTrigger, toTrigger, fromArc, toArc}] = cost;
    }

    // Get filename from path
    std::string name = filePath.substr(filePath.find_last_of("/\\") + 1);
    return std::make_unique<Instance>(N, edges, relations, name);
}

double Instance::computeObjective(const std::vector<int>& tour) const {
    std::vector<int> completeTour = tour;
    if (completeTour.back() != 0) {
        completeTour.push_back(0);
    }
    return static_cast<double>(computePartialTourCost(completeTour));
}

bool Instance::checkSolutionCorrectness(const std::vector<int>& tour) const {
    if (tour.back() == 0) {
        std::vector<int> tourCopy = tour;
        tourCopy.pop_back();
        return checkSolutionCorrectness(tourCopy);
    }

    if (tour.size() != static_cast<size_t>(N_) || tour[0] != 0) {
        return false;
    }

    boost::unordered_set<int> uniqueNodes(tour.begin(), tour.end());
    return uniqueNodes.size() == static_cast<size_t>(N_) &&
           std::all_of(uniqueNodes.begin(), uniqueNodes.end(), [this](int node) {
               return node >= 0 && node < N_;
           });
}

bool Instance::testSolution(const std::vector<int>& tour, double proposedObjective) const {
    if (!checkSolutionCorrectness(tour)) {
        return false;
    }

    double cost = computeObjective(tour);
    return std::abs(cost - proposedObjective) < 1e-6;
}

void Instance::saveSolution(const std::vector<int>& tour, std::optional<double> objective) {
    if (!checkSolutionCorrectness(tour)) {
        throw std::runtime_error("Invalid solution for instance " + name_);
    }

    double obj = objective.value_or(computeObjective(tour));
    if (!testSolution(tour, obj)) {
        std::cerr << "Warning: Solution does not have correct objective value for instance "
                  << name_ << " (computed: " << computeObjective(tour) << ", proposed: " << obj
                  << ")" << std::endl;
        obj = computeObjective(tour);
    }

    // Create solutions directory if it doesn't exist
    fs::path solutionsDir = fs::current_path() / "solutions";
    fs::create_directories(solutionsDir);

    // Get current timestamp
    auto now = std::chrono::system_clock::now();
    auto time = std::chrono::system_clock::to_time_t(now);
    std::stringstream ss;
    ss << std::put_time(std::localtime(&time), "%Y-%m-%d_%H-%M-%S");
    std::string timestamp = ss.str();

    // Write solution to file
    std::ofstream file(solutionsDir / name_, std::ios::app);
    if (file.tellp() != 0) {
        file << "\n";
    }

    for (size_t i = 0; i < tour.size(); ++i) {
        file << tour[i];
        if (i < tour.size() - 1)
            file << ",";
    }
    file << " | " << obj << " | " << timestamp;
}

std::optional<std::vector<int>> Instance::getBestKnownSolution(int idx) const {
    fs::path solutionPath = fs::path("solutions") / name_;
    if (!fs::exists(solutionPath)) {
        return std::nullopt;
    }

    std::vector<std::pair<std::string, double>> solutions;
    std::ifstream file(solutionPath);
    std::string line;
    while (std::getline(file, line)) {
        size_t pos = line.find(" | ");
        if (pos != std::string::npos) {
            std::string tour = line.substr(0, pos);
            double obj = std::stod(line.substr(pos + 3));
            if (std::find_if(solutions.begin(), solutions.end(), [&tour](const auto& s) {
                    return s.first == tour;
                }) == solutions.end()) {
                solutions.push_back({tour, obj});
            }
        }
    }

    std::sort(solutions.begin(), solutions.end(), [](const auto& a, const auto& b) {
        return a.second < b.second;
    });

    if (static_cast<size_t>(idx) >= solutions.size()) {
        throw std::runtime_error("Solution index out of range");
    }

    std::vector<int> tour;
    std::stringstream ss(solutions[idx].first);
    std::string item;
    while (std::getline(ss, item, ',')) {
        tour.push_back(std::stoi(item));
    }

    std::cout << "Best known solution for " << name_ << ": ";
    for (int node : tour) {
        std::cout << node << " ";
    }
    std::cout << "with objective " << solutions[idx].second << std::endl;

    return tour;
}

std::vector<boost::unordered_map<std::string, double>> Instance::getMipStart(
    bool useTspOnly) const {
    auto bestKnownSolution = getBestKnownSolution();
    std::vector<int> tour =
        bestKnownSolution && !useTspOnly ? bestKnownSolution.value() : tspSolution();
    return getVariablesFromTour(tour);
}

std::vector<boost::unordered_map<std::string, double>> Instance::getVariablesFromTour(
    const std::vector<int>& tour) const {
    // Create tour edges
    std::vector<std::pair<int, int>> tourEdges;
    for (size_t i = 0; i < tour.size(); ++i) {
        tourEdges.push_back({tour[i], (i < tour.size() - 1) ? tour[i + 1] : 0});
    }
    assert(tourEdges.size() == static_cast<size_t>(N_));

    // Initialize x variables (edge variables)
    boost::unordered_map<std::string, double> x;
    for (const auto& [edge, _] : edges_) {
        x[std::to_string(edge.first) + "," + std::to_string(edge.second)] =
            std::find(tourEdges.begin(), tourEdges.end(), edge) != tourEdges.end() ? 1.0 : 0.0;
    }

    // Initialize u variables (node position variables)
    boost::unordered_map<std::string, double> u;
    for (size_t i = 0; i < tour.size(); ++i) {
        u[std::to_string(tour[i])] = static_cast<double>(i);
    }

    // Initialize y variables (relation variables)
    boost::unordered_map<std::string, double> y;
    for (const auto& [rel, _] : relations_) {
        y[std::to_string(std::get<0>(rel)) + "," + std::to_string(std::get<1>(rel)) + "," +
          std::to_string(std::get<2>(rel)) + "," + std::to_string(std::get<3>(rel))] = 0.0;
    }

    // Set y variables based on triggering relations
    for (const auto& a : tourEdges) {
        assert(u[std::to_string(a.first)] < u[std::to_string(a.second)] || a.second == 0);

        auto it = R_a_.find(a);
        if (it == R_a_.end()) {
            continue;
        }

        // Find triggering relations that are in the tour
        std::vector<std::pair<int, int>> triggering;
        for (const auto& b : it->second) {
            if (std::find(tourEdges.begin(), tourEdges.end(), b) != tourEdges.end()) {
                triggering.push_back(b);
            }
        }

        if (triggering.empty()) {
            continue;
        }

        // Sort triggering by their index in tour_edges
        std::sort(
            triggering.begin(), triggering.end(), [&tourEdges](const auto& b1, const auto& b2) {
                return std::find(tourEdges.begin(), tourEdges.end(), b1) <
                       std::find(tourEdges.begin(), tourEdges.end(), b2);
            });

        // Remove triggering arcs that happen after arc a
        auto aPos = std::find(tourEdges.begin(), tourEdges.end(), a);
        triggering.erase(
            std::remove_if(triggering.begin(),
                           triggering.end(),
                           [&tourEdges, aPos](const auto& b) {
                               return std::find(tourEdges.begin(), tourEdges.end(), b) >= aPos;
                           }),
            triggering.end());

        if (!triggering.empty()) {
            auto lastTrigger = triggering.back();
            std::string key = std::to_string(lastTrigger.first) + "," +
                              std::to_string(lastTrigger.second) + "," + std::to_string(a.first) +
                              "," + std::to_string(a.second);
            y[key] = 1.0;
        }
    }

    // Initialize z variables (precedence variables)
    boost::unordered_map<std::string, double> z;
    for (const auto& [a1, a2, b1, b2] : zVarIndices_) {
        std::string key = std::to_string(a1) + "," + std::to_string(a2) + "," + std::to_string(b1) +
                          "," + std::to_string(b2);
        z[key] = u[std::to_string(a1)] + 1 <= u[std::to_string(b1)] ? 1.0 : 0.0;
    }

    return {x, y, u, z};
}

float Instance::computePartialTourCost(const std::vector<int>& partialTour, int startIdx) const {
    std::vector<std::pair<int, int>> path;
    for (size_t i = 0; i < partialTour.size() - 1; ++i) {
        path.push_back({partialTour[i], partialTour[i + 1]});
    }

    double cost = 0.0;
    // Add edge costs from startIdx to end of partialTour
    for (size_t i = startIdx; i < path.size(); ++i) {
        cost += edges_.at(path[i]);
    }

    // Handle relations from startIdx to end of partialTour
    for (auto aPos = path.begin() + startIdx; aPos != path.end(); ++aPos) {
        const auto& a = *aPos;
        auto it = R_a_.find(a);
        if (it == R_a_.end())
            continue;

        // Iterate backwards from the current position to find the last triggering edge
        for (auto edgeIt = aPos - 1; edgeIt >= path.begin(); --edgeIt) {
            const auto& b = *edgeIt;
            // Check if this edge is a potential trigger for edge a
            if (std::find(it->second.begin(), it->second.end(), b) != it->second.end()) {
                // Found the last trigger before the target edge
                cost += relations_.at({b.first, b.second, a.first, a.second});
                break;
            }
        }
    }

    return static_cast<float>(cost);
}

void Instance::generateZVarIndices() const {
    // Generate z variable indices
    boost::unordered_set<std::tuple<int, int, int, int>> zVarIndicesSet;
    for (const auto& [a, bList] : R_a_) {
        for (const auto& b : bList) {
            if (b != std::make_pair(0, 0)) {
                zVarIndicesSet.insert({a.first, a.second, b.first, b.second});
                zVarIndicesSet.insert({b.first, b.second, a.first, a.second});
            }
            for (const auto& c : bList) {
                if (b != c && b != std::make_pair(0, 0)) {
                    zVarIndicesSet.insert({b.first, b.second, c.first, c.second});
                }
            }
        }
    }
    zVarIndices_ =
        std::vector<std::tuple<int, int, int, int>>(zVarIndicesSet.begin(), zVarIndicesSet.end());
}

std::vector<int> Instance::tspSolution() const {
    GurobiTSPModel model(*this);
    model.formulate();
    model.solveToOptimality();
    return model.getBestTour();
}

void Instance::get_two_opt_neigbhor(std::vector<int>& tour) {
    // modify (inline) the tour passed as parameter 
    // Warning: Doesn't check solution validity
    int n = tour.size();

    srand(std::time(nullptr));
    int a = rand() % (n);
    int b = rand() % (n);
    
    std::swap(tour[a], tour[b]);
}

std::vector<std::vector<int>> Instance::get_all_two_opt_neigbhor(std::vector<int>& tour) {
    // returns all neighbors of the solution 
    // Warning: Doesn't check solution validity
    int n = tour.size();
    std::vector<std::vector<int>> res;

    for (int i = 0; i < n - 1; ++i) {
        for (int j = i + 1; j < n; ++j) {
            std::vector<int> neighbor = tour;
            std::swap(neighbor[i], neighbor[j]);
            res.push_back(neighbor);
        }
    }

    return res;
}